### PR TITLE
Liu x86 64

### PR DESCRIPTION
**Pull request by yaokai-liu**
Add x86_64 source files and workflow.

### CODE REVIEW

File: tests/testQuickSort.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+#define sqsort  qsort
+#define ELE_TYPE    int64_t
+#define BUFFER_LEN  0x800
+
+int cmp(const void *a, const void *b) {
+    return *(ELE_TYPE *)a > *(ELE_TYPE *)b ? 1 : -1;
+}
+
+TEST(TestQuickSort1, TestQuickSort) {
```
Comment (KKiiim) [inline tests/testQuickSort.cpp:16]: It's TOO complex to test the customer qsort implementation.
May be just need to compare whether every member is larger(smaller) than the next one.

**Review by KKiiim — CHANGES_REQUESTED** [review]

File: tests/testAbs.cpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#include"gtest/gtest.h"
+#include "functions.h"
+
+TEST(TestAbs32, TestAbs){
+    EXPECT_EQ(abs32(0), abs(0));
```
Comment (KKiiim) [inline tests/testAbs.cpp:5]: In test, need not to test `abs` in libc. Just expect(youfunc, 0)
Other test files are the same

File: include/functions.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#endif
+
+#include<stddef.h>
+#include<stdint.h>
+
+int32_t abs32(int32_t _num);
+int64_t abs64(int64_t _num);
+void * _memset(void * _dest, int _ch, size_t _count);
+int64_t getSum(int64_t * _nums, size_t _count);
+void cqsort(void * const _nums, size_t _count, size_t _ele_size,
```
[the remark this answers is not in the document]
Comment (yaokai-liu) [reply include/functions.h:16]: `cqsort` and `_sqort` are both aligned with standard C library function 
```
extern void qsort (void *__base, size_t __nmemb, size_t __size,
		   __compar_fn_t __compar) __nonnull ((1, 4));
```
in `stdlib.h`.
And implement within template is too hard for assembly.

File: tests/testGetSum.cpp [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#include"gtest/gtest.h"
+#include "functions.h"
+
+int64_t sum(int64_t * p, size_t c) {
+    int64_t s = 0;
+    for (size_t i = 0; i < c; i ++) { s += p[i]; }
+    return s;
```
[the remark this answers is not in the document]
Comment (yaokai-liu) [reply tests/testGetSum.cpp:7]: Maybe the value of `s` would be overflow out the max of  `int64_t`, but I think that should be handled by one who calls it.

[the remark this answers is not in the document]
Comment (KKiiim) [reply tests/testGetSum.cpp:7]: Of course the caller should understand . But if the return value is invalid, how can the caller know that the answer is wrong.

File: csrc/cqsort.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#endif
+#include "functions.h"
+#include <alloca.h>
+
+void cqsort(void * const _nums, size_t _count, size_t _ele_size, int (*_fn_cmp)(const void *, const void *)) {
+    if (_count < 1) { return ; }
+    char * lower_bound = (char *)_nums;
+    char * upper_bound = (char *)_nums + (_count - 1) * _ele_size;
+    char *left=lower_bound, *right=upper_bound;
+    char * base = (char *)alloca(_ele_size);
```
[the remark this answers is not in the document]
Comment (KKiiim) [reply csrc/cqsort.cpp:20]: refactor and delete the multi `alloca`

File: tests/testGetSum.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (KKiiim) [reply tests/testGetSum.cpp:7]: Which means, your api is unsafe and confused

File: tests/testQuickSort.cpp (hunk 1 above)
Comment (yaokai-liu) [reply tests/testQuickSort.cpp:16]: an equivalent implementation of `int (*__compar_fn_t) (const void *, const void *)` which requested by `qsort` is like
```
int cmp (const void * a, const void *b) {
    return *(int64_t *)a - *(int64_t *)b;
};
```

Actually, any function with signature `int (*__compar_fn_t) (const void *, const void *)` and always return negative value while first param less than second param could be the `cmp`.

Comment (KKiiim) [reply tests/testQuickSort.cpp:16]: What you're saying is nothing to do with this comment

File: tests/testQuickSort.cpp [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+TEST(TestQuickSortReservedInt64Array, TestQuickSort) {
+    auto * buffer1 = (ELE_TYPE *) malloc(BUFFER_LEN * sizeof(ELE_TYPE));
+    auto * buffer2 = (ELE_TYPE *) malloc(BUFFER_LEN * sizeof(ELE_TYPE));
+    auto * buffer3 = (ELE_TYPE *) malloc(BUFFER_LEN * sizeof(ELE_TYPE));
+
+    srandom(time(nullptr));
+    int8_t shift = (int8_t) random() % 64;
+    for (int i = 0; i < BUFFER_LEN; i++){ buffer1[i] = BUFFER_LEN - i; buffer1[i] << shift; }
```
Comment (KKiiim) [inline tests/testQuickSort.cpp:68]: Does `buffer1[i] << shift` change any member's value?

File: tests/testQuickSort.cpp [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+TEST(TestQuickSortConstInt64Array, TestQuickSort) {
+    auto * buffer1 = (ELE_TYPE *) malloc(BUFFER_LEN * sizeof(ELE_TYPE));
+    auto * buffer2 = (ELE_TYPE *) malloc(BUFFER_LEN * sizeof(ELE_TYPE));
+    auto * buffer3 = (ELE_TYPE *) malloc(BUFFER_LEN * sizeof(ELE_TYPE));
+
+    srandom(time(nullptr));
+    int64_t value = random();
+    for (int i = 0; i < BUFFER_LEN; i++){ buffer1[i] = value; }
```
Comment (KKiiim) [inline tests/testQuickSort.cpp:46]: Why should test qsort of an array which has all members the same.

**Review by KKiiim — CHANGES_REQUESTED** [review]

File: tests/testQuickSort.cpp (hunk 7 above)
Comment (yaokai-liu) [reply tests/testQuickSort.cpp:46]: To test an array all elements are the same.

File: tests/testGetSum.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (yaokai-liu) [reply tests/testGetSum.cpp:7]: `cqsort` has been refactored with `cswap`.

File: tests/testQuickSort.cpp [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    _qsort(buffer3, BUFFER_LEN, sizeof(ELE_TYPE), cmp64);
+    EXPECT_EQ(memcmp(buffer2, buffer1, BUFFER_LEN), 0);
+    EXPECT_EQ(memcmp(buffer3, buffer1, BUFFER_LEN), 0);
+
+    free(buffer1);
+    free(buffer2);
+    free(buffer3);
+}
+
+TEST(TestQuickSortReverseInt64Array, TestQuickSort) {
```
Comment (KKiiim) [inline tests/testQuickSort.cpp:61]: `random()` is a random number. Excluding meaningless performance expenses, what's the point of using `shift` for double randomization.

**Review by KKiiim — APPROVED** [review]

**Pull request merged**